libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -567,7 +565,68 @@ internal object MapboxRouteLineUtils {
                     else -> true
                 }
             }
-        }.cacheResult(extractRouteDataCacheDeDuped)
+        }
+
+    /**
+     * Extracts restricted road section data from the route. Most routes do not contain restricted
+     * sections. This can be an expensive operation.The implementation defers or avoids the most
+     * expensive operations as much as possible.
+     */
+    internal val extractRouteRestrictionData: (
+        route: NavigationRoute,
+    ) -> List<ExtractedRouteRestrictionData> =
+        { route: NavigationRoute ->
+            val itemsToReturn = mutableListOf<ExtractedRouteRestrictionData>()
+            val granularDistances by lazy {
+                granularDistancesProvider(route)
+            }
+            route.directionsRoute.legs()?.forEachIndexed { legIndex, leg ->
+                val intersections = leg.steps()?.map {
+                    it.intersections() ?: emptyList()
+                }?.flatten()
+                val filteredIntersections =
+                    intersections?.filterIndexed { index, stepIntersection ->
+                        val isRestricted =
+                            stepIntersection.classes()?.contains("restricted") ?: false
+                        val previousIsRestricted = index != 0 &&
+                            intersections[index - 1].classes()?.contains("restricted") ?: false
+                        (isRestricted || index == 0 || previousIsRestricted)
+                    }
+                ifNonNull(filteredIntersections) { stepIntersections ->
+                    val legDistancesArray =
+                        granularDistances?.nestedDistancesArray?.map { legDistances ->
+                            legDistances.flatten()
+                        }
+                    stepIntersections.forEachIndexed { intersectionIndex, stepIntersection ->
+                        val isRestricted = stepIntersection.classes()?.contains("restricted") ?: false
+                        val previousIsRestricted = intersectionIndex != 0 &&
+                            stepIntersections[intersectionIndex - 1].classes()
+                                ?.contains("restricted") ?: false
```

Comment:
We're iterating over `filteredIntersections`, won't all of the return `true` here? This would break the logic of injecting a non-restricted intersection if the previous one was restricted which is needed to create valid expressions.

Reply:
Maybe we can make `filteredIntersections` a richer data structure to include original indices, or find another way of injecting a non-restricted intersections.

Reply:
Hmm, I think I see that we're allowing non-restricted intersections to land in the `filteredIntersections` collection as well.

In this case we can try making the collection built of:
```
data class FilteredIntersection(intersection, isRestricted)
```
so that we don't have to check for it twice.

Reply:
I made some adjustments.